sweethome3d impure? copied package don't work

**Describe the bug**
I copied a closure from one system (19.03) to a USB thumb drive and from that to another system (19.09), but the program don't work.

I think nix packages should be pure and work also on another system running a different NixOS channel. (Right?)

**To Reproduce**
Steps to reproduce the behavior:
1. have sweethome3d installed
2. plug in thumb drive
3. mount thumb drive (i use pantheons files app and open a terminal from there, then `pwd`)
4. create a folder on the thumb drive. mine is "nixpkgs"
5. copy to that folder: `nix copy --to file:///run/media/davidak/9AEE-1F0D/nixpkgs $(type -p sweethome3d)`
6. get the nix store path of the program:

```
[davidak@X230:~]$ type -p sweethome3d
/home/davidak/.nix-profile/bin/sweethome3d

[davidak@X230:~]$ ls -la /home/davidak/.nix-profile/bin/sweethome3d
lrwxrwxrwx 1 root root 87  1. Jan 1970  /home/davidak/.nix-profile/bin/sweethome3d -> /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d
```
7. plug the usb drive into another system and mount it, get the path
8. copy the nix store path to your system: `nix copy --from file:///run/media/davidak/9AEE-1F0D/nixpkgs /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
9.a you can install it for the current user if you like: `nix-env -i /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
9.b or just open a nix shell: `nix run /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
10. now run the program:

```
$ sweethome3d 
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.eteks.sweethome3d.SweetHome3DBootstrap.main(Unknown Source)
Caused by: java.lang.UnsatisfiedLinkError: /tmp/extension4577508567935158164.so: libXxf86vm.so.1: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at com.jogamp.common.jvm.JNILibLoaderBase.loadLibraryInternal(JNILibLoaderBase.java:603)
	at com.jogamp.common.jvm.JNILibLoaderBase.access$000(JNILibLoaderBase.java:63)
	at com.jogamp.common.jvm.JNILibLoaderBase$DefaultAction.loadLibrary(JNILibLoaderBase.java:106)
	at com.jogamp.common.jvm.JNILibLoaderBase.loadLibrary(JNILibLoaderBase.java:487)
	at jogamp.nativewindow.NWJNILibLoader.access$000(NWJNILibLoader.java:39)
	at jogamp.nativewindow.NWJNILibLoader$1.run(NWJNILibLoader.java:49)
	at jogamp.nativewindow.NWJNILibLoader$1.run(NWJNILibLoader.java:41)
	at java.security.AccessController.doPrivileged(Native Method)
	at jogamp.nativewindow.NWJNILibLoader.loadNativeWindow(NWJNILibLoader.java:41)
	at jogamp.nativewindow.jawt.JAWTUtil.<clinit>(JAWTUtil.java:336)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.jogamp.nativewindow.NativeWindowFactory$3.run(NativeWindowFactory.java:344)
	at com.jogamp.nativewindow.NativeWindowFactory$3.run(NativeWindowFactory.java:340)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.jogamp.nativewindow.NativeWindowFactory.initSingleton(NativeWindowFactory.java:340)
	at com.jogamp.opengl.GLProfile.initProfilesForDefaultDevices(GLProfile.java:1721)
	at com.jogamp.opengl.GLProfile.access$000(GLProfile.java:80)
	at com.jogamp.opengl.GLProfile$1.run(GLProfile.java:230)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.jogamp.opengl.GLProfile.initSingleton(GLProfile.java:216)
	at com.jogamp.opengl.GLProfile.getProfileMap(GLProfile.java:2297)
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:1021)
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:1050)
	at com.jogamp.opengl.GLProfile.getMaxFixedFunc(GLProfile.java:803)
	at javax.media.j3d.JoglPipeline.initialize(JoglPipeline.java:131)
	at javax.media.j3d.Pipeline.createPipeline(Pipeline.java:92)
	at javax.media.j3d.MasterControl.loadLibraries(MasterControl.java:858)
	at javax.media.j3d.VirtualUniverse.<clinit>(VirtualUniverse.java:267)
	at javax.media.j3d.GraphicsConfigTemplate3D.isGraphicsConfigSupported(GraphicsConfigTemplate3D.java:348)
	at com.eteks.sweethome3d.j3d.Component3DManager.createGraphicsConfigurationTemplate3D(Unknown Source)
	at com.eteks.sweethome3d.j3d.Component3DManager.<init>(Unknown Source)
	at com.eteks.sweethome3d.j3d.Component3DManager.getInstance(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.addComponent3DRenderingErrorObserver(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.init(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.main(Unknown Source)
	... 5 more
^C
```

**Expected behavior**
Program just work.

**Additional context**
I moved to a new flat with most of my computers and want to plan where to put my furniture. I installed sweethome3d on my notebook, but it is not powerfull enough. So i want to copy the program to my iMac. I have internet from my tethered mobile phone, so i can't just install it from the internet! sweethome3d is about 500 MB, that would be 1/4 of my data plan!!!

**Metadata**
X230:
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.63, NixOS, 19.03.173221.9ca57dc9171 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(davidak): `""`
 - channels(root): `"nixos-19.03.173221.9ca57dc9171, nixos-hardware, nixos-unstable-19.09pre186857.239fffc90d7"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

imac:
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.63, NixOS, 19.09.git.7db3602 (Loris)`
hangs after that

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: sweethome3d
# a list of nixos modules affected by the problem
module:
```

When i run the package from unstable on stable, i get the same error:

```
[davidak@X230:~]$ nix run -f channel:nixos-unstable sweethome3d.application
[113 copied (464.8 MiB), 139.9 MiB DL]

[davidak@X230:~]$ sweethome3d 
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.eteks.sweethome3d.SweetHome3DBootstrap.main(Unknown Source)
Caused by: java.lang.UnsatisfiedLinkError: /tmp/extension8491416013782226237.so: libXxf86vm.so.1: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
...
```
(but it will hopefully work on the imac when i copy it there again)

--- a/pkgs/applications/misc/sweethome3d/default.nix
+++ b/pkgs/applications/misc/sweethome3d/default.nix
@@ -10,6 +10,7 @@
 , gtk3
 , gsettings-desktop-schemas
 , p7zip
+, autoPatchelfHook
 , libXxf86vm
 , unzip
 }:
@@ -41,14 +42,17 @@ let
     };
 
     postPatch = ''
-      patchelf --set-rpath ${libXxf86vm}/lib lib/java3d-1.6/linux/amd64/libnativewindow_awt.so
-      patchelf --set-rpath ${libXxf86vm}/lib lib/java3d-1.6/linux/amd64/libnativewindow_x11.so
-      patchelf --set-rpath ${libXxf86vm}/lib lib/java3d-1.6/linux/i586/libnativewindow_awt.so
-      patchelf --set-rpath ${libXxf86vm}/lib lib/java3d-1.6/linux/i586/libnativewindow_x11.so
+      addAutoPatchelfSearchPath ${jre8}/lib/openjdk/jre/lib/
+      autoPatchelf lib
+
+      # Nix cannot see the runtime references to the paths we just patched in
+      # once they've been compressed into the .jar. Scan for and remember them
+      # as plain text so they don't get overlooked.
+      find . -name '*.so' | xargs strings | { grep '/nix/store' || :; } >> ./.jar-paths
     '';
 
-    nativeBuildInputs = [ makeWrapper unzip ];
-    buildInputs = [ ant jdk8 p7zip gtk3 gsettings-desktop-schemas ];
+    nativeBuildInputs = [ makeWrapper unzip autoPatchelfHook ];
+    buildInputs = [ ant jdk8 p7zip gtk3 gsettings-desktop-schemas libXxf86vm ];
 
     buildPhase = ''
       runHook preBuild
@@ -79,6 +83,13 @@ let
         --set MESA_GL_VERSION_OVERRIDE 2.1 \
         --prefix XDG_DATA_DIRS : "$XDG_ICON_DIRS:${gtk3.out}/share:${gsettings-desktop-schemas}/share:$out/share:$GSETTINGS_SCHEMAS_PATH" \
         --add-flags "-Dsun.java2d.opengl=true -jar $out/share/java/${module}-${version}.jar -cp $out/share/java/Furniture.jar:$out/share/java/Textures.jar:$out/share/java/Help.jar -d${toString stdenv.hostPlatform.parsed.cpu.bits}"
+
+
+      # remember the store paths found inside the .jar libraries. note that
+      # which file they are in does not matter in particular, just that some
+      # file somewhere lists them in plain-text
+      mkdir -p $out/nix-support
+      cp .jar-paths $out/nix-support/depends
 
       runHook postInstall
     '';

